### PR TITLE
upstream: implement Cluster's load_assignment field

### LINKED ISSUE #439: Allow healthcheck on different port for non-EDS cluster types
The http healthcheck will allow a path but not a port, which is problematic for existing Dropwizard-based services which by default to have the healthcheck endpoint on a separate admin port. There are also instances of proxies where a tcp check is being done to check the health of the proxy but the traffic is going over different ports.

Envoy would need a healthcheck_port in the static configuration and the SDS API to support being a drop-in replacement for HAProxy under these conditions.

http://www.dropwizard.io/1.0.6/docs/manual/core.html#health-checks

### PR DESCRIPTION
**Pull request by dio**
*Description*:
This patch implements `load_assigment` field in CDS' `Cluster`.
This change specifically adds the implementation of the new `load_assigment` field
for clusters with discovery-type: `STATIC`, `STRICT_DNS` and `LOGICAL_DNS`.

While adding this `load_assigment` field implementation to `Cluster`, 
this patch also allows specifying optional (active) health check config per specified upstream host.

*Risk Level*: medium
*Testing*: unit tests
*Docs Changes*:
- This unhides docs for endpoint health check config

*Release Notes*: N/A

Fixes #439

Signed-off-by: Dhi Aurrahman <dio@rockybars.com>

### CODE REVIEW

**Comment by dio** [comment]
cc. @danielhochman

File: docs/root/configuration/overview/v2_overview.rst [hunk 1]
```diff
@@ -544,10 +565,10 @@ Management Server Unreachability
 --------------------------------
 
 When Envoy instance looses connectivity with the management server, Envoy will latch on to
```
Comment (danielhochman) [inline docs/root/configuration/overview/v2_overview.rst]: while you're here, do you mind `s/looses/loses/`

File: docs/root/intro/arch_overview/health_checking.rst [hunk 2]
```diff
@@ -24,6 +24,37 @@ unhealthy, successes required before marking a host healthy, etc.):
...
   :ref:`redis_key <config_cluster_manager_cluster_hc_redis_key>`.
 
+.. _arch_overview_per_cluster_health_check_config:
+
+Per cluster member health check config
+--------------------------------------
+
+If active health checking is configured for an upstream cluster, a specific additional configuration
+for each registered member can be specified by setting the
+:ref:`health check config<envoy_api_msg_endpoint.Endpoint.HealthCheckConfig>`
```
Comment (danielhochman) [inline docs/root/intro/arch_overview/health_checking.rst]: nit: can we keep the CamelCase for `HealthCheckConfig` and `Endpoint`? or at least make it consistent for each of the linked resources in this paragraph.

File: docs/root/intro/arch_overview/health_checking.rst [hunk 3]
```diff
@@ -24,6 +24,37 @@ unhealthy, successes required before marking a host healthy, etc.):
...
+in :ref:`endpoint message<envoy_api_msg_endpoint.Endpoint>`
```
Comment (danielhochman) [inline docs/root/intro/arch_overview/health_checking.rst]: slight rewording, add `the` after `in`

File: docs/root/intro/arch_overview/health_checking.rst [hunk 4]
```diff
@@ -24,6 +24,37 @@ unhealthy, successes required before marking a host healthy, etc.):
...
+
+.. code-block:: yaml
+
+  load_assignment:
+    endpoints:
+    - lb_endpoints:
+      - endpoint:
+          health_check_config:
+            port_value: 8080
+
```
Comment (danielhochman) [inline docs/root/intro/arch_overview/health_checking.rst]: nit: remove newline

**Comment by dio** [comment]
Thanks, @danielhochman. Updated in 18e98f4.

**Comment by dio** [comment]
Since this is a re-submitted PR of #3261, probably this require @mattklein123 @alyssawilk @htuch and @zuercher to check too.

File: docs/root/configuration/overview/v2_overview.rst [hunk 5]
```diff
@@ -543,11 +564,11 @@ the shared ADS channel.
 Management Server Unreachability
 --------------------------------
 
-When Envoy instance looses connectivity with the management server, Envoy will latch on to
-the previous configuration while actively retrying in the background to reestablish the 
-connection with the management server. 
+When Envoy instance loses connectivity with the management server, Envoy will latch on to
```
Comment (zuercher) [inline docs/root/configuration/overview/v2_overview.rst]: While we're here, it should read "When an Envoy instance loses..."

File: source/common/upstream/logical_dns_cluster.cc [hunk 6]
```diff
@@ -27,10 +28,17 @@ LogicalDnsCluster::LogicalDnsCluster(const envoy::api::v2::Cluster& cluster,
...
-  if (hosts.size() != 1) {
-    throw EnvoyException("logical_dns clusters must have a single host");
+      resolve_timer_(dispatcher.createTimer([this]() -> void { startResolve(); })),
+      local_info_(local_info),
+      load_assignment_(cluster.has_load_assignment()
+                           ? cluster.load_assignment()
+                           : Config::Utility::translateClusterHosts(cluster.hosts())) {
+  const auto& locality_lb_endpoints = load_assignment_.endpoints();
+  if (locality_lb_endpoints.size() != 1 || locality_lb_endpoints[0].lb_endpoints().size() != 1) {
+    throw EnvoyException(fmt::format("LOGICAL_DNS clusters must have {}",
```
Comment (zuercher) [inline source/common/upstream/logical_dns_cluster.cc]: nit: just `if (cluster.has_load_assignment()) { /* throw exception message1 */ } else { /* throw exception message 2 */ }`

(it'll be easier to search for the error string, but I don't feel strongly)

File: docs/root/configuration/overview/v2_overview.rst (hunk 5 above)
Comment (dio) [reply docs/root/configuration/overview/v2_overview.rst]: Thanks, @zuercher! This is addressed in 75a1537.

File: source/common/upstream/logical_dns_cluster.cc (hunk 6 above)
Comment (dio) [reply source/common/upstream/logical_dns_cluster.cc]: :+1: updated 75a1537.

**Pull request merged**